### ISSUE TITLE
Not getting to MCU select - Sublime Text 3

### ISSUE THREAD

**Issue by Gasman2014** [kblomqvist/SublimeAVR#9]
Very promising plug in but..

OS X (10.11), Sublime Text 3 (3103), AVR

I have installed CrossPack (path /usr/local/CrossPack-AVR/bin/avr-gcc) and adjusted the path in my user settings to reflect this. `print(target)` in unix.py confirms the correct path is being used. When I run the 'AVR' command from the command palette it initially identifies the missing 'clang' and downloads it. It does not get as far as 'Select MCU'. The result of `/usr/local/CrossPack-AVR/bin/avr-as -mlist-devices` is a long list of MCU types. Printing variables within SublimeAVR.py show that `print(self.initial_location)` is ~ (as expected) but `print(self.devices)` is  [ ]. Don't think the code gets beyond here...

Not sure how to get any further than this (I know very little python!) but any help would be appreciated!.It sounds as though this was similar to an issue another user had but I don't think that was resolved. I have tried downgrading to ST2 but exactly the same problem occurs.

John

**Comment by Gasman2014** [comment]
I realise that this is part of an older issue with 'cwd' - having looked through all the previous commits. Clearly it is a bit of a problem - but despite a bit of tinkering, I have not made much progress. Would really like to get this working though. One observation - I have 2 versions of avr-as - 2.23 installed from a CrossPack install and 2.25 from a Brew install (don't ask why!). `avr-as -mlist-devices` both return a list of MCU types but the 2.25 version also reports '`Fatal error: unknown MCU: list-devices`'  - presume this is actually an avr-as error? Not sure if this is significant as the current version stalls before returning an mcc list.

**Comment by kblomqvist** [comment]
Yep that's the case. I checked that `subprocess.call` accepts 'cwd' as a parameter too (though not documented in reference guide)... Maybe using `call` instead of `popen` could work. However, I do not have mac to test it.

Regarding to `avr-as`, does both versions list the devices but with 2.25 you also get an error `Fatal error: unknown MCU: list-devices` into 'stderr' although the list in 'stdout' is ok?

**Comment by Gasman2014** [comment]
I have tried using `call` instead of `popen` but no luck. 
I directed `/usr/local/bin/avr-as -mlist-devices 2 >` (avr-as 2.25) and the entire MCU list and fatal error is sent to stderr whereas with `/usr/local/CrossPack-AVR/bin/avr-as` (version avr-as 2.23) lists the mcus in stdout and no output in stderr. This second behaviour on the older version would seem to be the more correct one!

**Comment by kblomqvist** [comment]
Seems like Atmel screwed up the avr-as device listing then in the newest version.

The issue related to 'cwd' you could try this:
```python
	p = subprocess.Popen(
		os.path.join(location, "avr-as") + " -mlist-devices",
		shell  = True,
		#cwd    = location, # This ain't working on OSX. Thus os.path.join() is used above.
		stdout = subprocess.PIPE,
		stderr = subprocess.PIPE
	)
	out, err = p.communicate()
```

**Comment by Gasman2014** [comment]
Yes! That works on OS X. Thank you so much - looks really promising. The package now runs and sets up a project based on the supplied template. At present it will not build because the mcu selected does not get copied into the makefile. 
```
TARGET=a
MMCU?=
AVRDUDE_FLAGS?=
```
It is fairly simple to add a 'atmega328p' or whatever manually but the intent is that this should be copied automatically. Shall I leave this in this thread or would you like me to flag this as a separate issue? It is, however, now a functional system - so many thanks.

**Comment by kblomqvist** [comment]
The MMCU should get copied into Makefile along with AVRDUDE_FLAGS.

**Comment by kblomqvist** [comment]
There might have been another issue with the fetch of avr-gcc predefines too. Because, I wasn't using a list but a string for `args` argument of Popen.

Could you verify if the latest commit f7777152e01635f77732795fda100f60777b823f still works for you?

**Comment by kblomqvist** [comment]
While testing the plugin (after years in Linux) I faced a problem with avr-gcc's `-x <language>` option, which prevented the creation of `.sublime-project` file. Did the plugin create this file for you?

Also seems like environment settings from `.sublime-project` don't work in Linux, so MMCU and AVRDUDE_FLAGS aren't propagated into Makefile.

**Comment by Gasman2014** [comment]
With regard to previous messages:

Re commit `f777715 `. This works to the extent that it creates a project in the selected path, as before. However, AVRDUDE_FLAGS and MMCU are still **not** being set in the makefile. 

However, in `f777715` the .sublime-project file is now **not** being created. In a previous version, it was created and contained the MMCU / AVRDUDE_FLAGS however, these were not being copied into the makefile. I think this was from the master branch at the time that you made the suggested alteration above i.e.

```
    p = subprocess.Popen(
        os.path.join(location, "avr-as") + " -mlist-devices",
        shell  = True,
        #cwd    = location, # This ain't working on OSX. Thus os.path.join() is used above.
        stdout = subprocess.PIPE,
        stderr = subprocess.PIPE
    )
    out, err = p.communicate()
```

Thanks for continuing to look at this.

**Comment by kblomqvist** [comment]
The idea is not to copy the MMCU and AVRDUDE_FLAGS into Makefile but propagate these via environment variables defined in `.sublime-project file`. Thus using `?=` operator in Makefile. Otherwise Makefile should be "templated".

I think I would need a Mac to really understand what's wrong. In Linux the plugin works now (the current master). It was broken when I started to test and I had to relearn how SublimeClang needs to be set up (now documented in readme).
